Assets/src/Components/Board/BoardMotion.cs

```diff
@@ -0,0 +1,342 @@
+            if (isWallClosed)
+            {
+                return;
+            }
+
+            if (stepData.Number == 1)
+            {
+                RemoveAllSteps();
+                RemoveTarget();
+
+                AddNewStep(
+                    position.ColumnPosition,
+                    position.RowPosition,
+                    stepData.StartNumber + 2,
+                    stepData.StartNumber + 1,
+                    direction
+                );
+                GenerateTarget(Target);
+
+                AddNewWall();
+            }
+            else
+            {
+                var isTargetReached = IsTargetReached(
+                    position.ColumnPosition,
+                    position.RowPosition,
+                    stepData.Number,
+                    direction
+                );
+                if (isTargetReached)
+                {
+                    var targetValue = GetTargetValue();
+                    RemoveAllSteps();
+                    RemoveTarget();
+
+                    AddNewStep(
+                        position.ColumnPosition,
+                        position.RowPosition,
+                        stepData.StartNumber + 2 + targetValue,
+                        stepData.StartNumber + 1 + targetValue,
+                        direction
+                    );
+                    GenerateTarget(Target);
+                }
+                else
+                {
+                    var isNextBusy = IsNextPositionBusy(
+                        position.ColumnPosition,
+                        position.RowPosition,
+                        direction
+                    );
+                    if (isNextBusy)
+                    {
+                        return;
+                    }
+                    AddNewStep(
+                        position.ColumnPosition,
+                        position.RowPosition,
+                        stepData.Number,
+                        stepData.StartNumber,
+                        direction
+                    );
+                }
+            }
+        }
+    }
+
+    private DirectionEnum InputToDirection()
+    {
+        DirectionEnum direction = DirectionEnum.None;
+        if (Input.GetKeyDown(KeyCode.LeftArrow))
+        {
+            direction = DirectionEnum.Left;
+        }
+        else if (Input.GetKeyDown(KeyCode.RightArrow))
+        {
+            direction = DirectionEnum.Right;
+        }
+        else if (Input.GetKeyDown(KeyCode.UpArrow))
+        {
+            direction = DirectionEnum.Top;
+        }
+        else if (Input.GetKeyDown(KeyCode.DownArrow))
+        {
+            direction = DirectionEnum.Bottom;
+        }
+
+        return direction;
+    }
+
+    private SnakeStepData GetLastSnakeStepData()
+    {
+        return GetComponentsInChildren<SnakeStepData>()
+            .OrderBy(data => data.Number)
+            .First();
+    }
+
+    private BoardElement GetBoardElementAtPosition(int column, int row)
```

Comment:
Separate wall behaviour from sneak behavior.  E.g. hang generating of wall on board and left here only sneak motion controller and use raycast to handle obstacles